In [25]:
import json
from pymongo import MongoClient

# Load secrets.json safely
with open('secrets.json') as f:
    secrets = json.load(f)


In [26]:
# Extract the MongoDB connection string
mongo_connection_string = secrets['mongodb']

# Initialize MongoDB client using the connection string
client = MongoClient(mongo_connection_string)

# Access the 'Data320' database and 'IMDB(flawed)' collection
db = client['Data320']
imdb_collection = db['IMDB(flawed)']

# Fetch a sample of movies to confirm the connection
sample_movies = list(imdb_collection.find().limit(5))
for movie in sample_movies:
    print(movie)

# Safe output: Display the length of the connection string
print(f"The MongoDB connection string is {len(mongo_connection_string)} characters long.")

The MongoDB connection string is 75 characters long.


In [27]:
# Connect to the 'IMDB' database (corrected)
my_database = client['IMDB']

# List all collections within the 'IMDB' database
all_collections = my_database.list_collection_names()
print(f"This database has the collections: {all_collections}")

This database has the collections: ['Metacritic', 'IMDB(flawed)']


In [28]:
# Connect to the 'IMDB(flawed)' collection
imdb_collection = my_database['IMDB(flawed)']

# Fetch a sample of movies to confirm the connection
sample_movies = list(imdb_collection.find().limit(5))

# Display the sample movies
for movie in sample_movies:
    print(movie)

{'_id': ObjectId('670dd57c93c8bf25ee148e95'), 'id': 77631, 'title': 'Grease', 'runtime': 110, 'user_rating': 7.2, 'votes': 265183, 'mpaa_rating': 'TV-14::(D)', 'release_date': '6/13/1978', 'budget': 6000000, 'opening_weekend': 60759, 'gross_sales': 394955690, 'genres': 'Comedy, Musical, Romance', 'cast': 'John Travolta, Olivia Newton-John, Stockard Channing, Jeff Conaway, Barry Pearl, Michael Tucci, Kelly Ward, Didi Conn, Jamie Donnelly, Dinah Manoff', 'director': 'Randal Kleiser', 'producer': 'Allan Carr, Neil A. Machlis, Robert Stigwood', 'company': 'Paramount Pictures, Robert Stigwood Organization, Allan Carr Production'}
{'_id': ObjectId('670dd57c93c8bf25ee148e96'), 'id': 78346, 'title': 'Superman', 'runtime': 143, 'user_rating': 7.4, 'votes': 172769, 'mpaa_rating': 'TV-PG::(LV)', 'release_date': '12/10/1978', 'budget': 55000000, 'opening_weekend': 7465343, 'gross_sales': 166200000, 'genres': 'Action, Adventure, Sci-Fi', 'cast': "Marlon Brando, Gene Hackman, Christopher Reeve, Ned 

In [29]:
import pandas as pd

# Fetch all documents from 'IMDB(flawed)' and convert to DataFrame
cursor = imdb_collection.find()
imdb_df = pd.DataFrame(cursor)

# Display the first few rows to confirm data retrieval
print(imdb_df.head())

  _id                        id    title                             runtime  user_rating  votes  mpaa_rating                release_date  budget      opening_weekend  gross_sales genres                     cast                                               director        producer                                           company                                           
0  670dd57c93c8bf25ee148e95  77631                           Grease  110.0    7.2          265183                 TV-14::(D)   6/13/1978    6000000.0    60759.0        394955690.0   Comedy, Musical, Romance  John Travolta, Olivia Newton-John, Stockard Ch...  Randal Kleiser       Allan Carr, Neil A. Machlis, Robert Stigwood  Paramount Pictures, Robert Stigwood Organizati...
1  670dd57c93c8bf25ee148e96  78346                         Superman  143.0    7.4          172769                TV-PG::(LV)  12/10/1978   55000000.0  7465343.0        166200000.0  Action, Adventure, Sci-Fi  Marlon Brando, Gene Hackman, Christopher

In [30]:
import re
import pandas as pd

# Specify the year to filter
year = 2010

# Query the collection for movies from the specified year
movies_2010 = list(imdb_collection.find({'release_date': re.compile(str(year))}))

# Check if movies were found
if not movies_2010:
    print("No movies found for the specified year.")


In [31]:
# Convert 'release_date' to datetime
imdb_2010_df.loc[:, 'release_date'] = pd.to_datetime(imdb_2010_df['release_date'], errors='coerce')

# Convert 'runtime' and 'budget' to numeric
imdb_2010_df.loc[:, 'runtime'] = pd.to_numeric(imdb_2010_df['runtime'], errors='coerce')
imdb_2010_df.loc[:, 'budget'] = pd.to_numeric(imdb_2010_df['budget'], errors='coerce')

# Display the first few rows to confirm data cleansing
print(imdb_2010_df.head())

  _id                        id      title                                          runtime  user_rating  votes   mpaa_rating               release_date  budget       opening_weekend  gross_sales  genres                                         cast                                               director           producer                                           company                                           
0  670dd58593c8bf25ee14ca20  1196141                          Diary of a Wimpy Kid   94      6.2            49261                TV-PG::(V) 2010-03-18     15000000.0   22126166.0      7.570050e+07                          Comedy, Drama, Family  Zachary Gordon, Robert Capron, Rachael Harris,...   Thor Freudenthal  Nina Jacobson, Jeff Kinney, Brad Simpson, Etha...  Color Force, Dayday Films, Dune Entertainment III
1  670dd58593c8bf25ee14ca21  1375666                                     Inception  148      8.8          2282266  TV-14::(DLV, TV Rating.) 2010-07-08    160000000.0   

In [32]:
# Sort by 'runtime' (ascending) to find the lowest runtime movies
lowest_runtime = imdb_2010_df.sort_values(by='runtime').head()
print("Movies with the Lowest Runtime in 2010:")
print(lowest_runtime[['title', 'runtime', 'budget', 'release_date']])

# Sort by 'runtime' (descending) to find the highest runtime movies
highest_runtime = imdb_2010_df.sort_values(by='runtime', ascending=False).head()
print("\nMovies with the Highest Runtime in 2010:")
print(highest_runtime[['title', 'runtime', 'budget', 'release_date']])

# Sort by 'budget' (ascending) to find the lowest budget movies
lowest_budget = imdb_2010_df.sort_values(by='budget').head()
print("\nMovies with the Lowest Budget in 2010:")
print(lowest_budget[['title', 'runtime', 'budget', 'release_date']])

# Sort by 'budget' (descending) to find the highest budget movies
highest_budget = imdb_2010_df.sort_values(by='budget', ascending=False).head()
print("\nMovies with the Highest Budget in 2010:")
print(highest_budget[['title', 'runtime', 'budget', 'release_date']])

Movies with the Lowest Runtime in 2010:
     title                             runtime  budget release_date
727                            Hubble  45      NaN     2010-03-19  
988                       Life Cycles  47      NaN     2010-09-21  
637       Yu-Gi-Oh! Bonds Beyond Time  49      NaN     2010-01-23  
1028  The Lives of Mount Druitt Youth  60      NaN     2010-08-10  
961            1915 Armenian Genocide  60      NaN     2010-10-22  

Movies with the Highest Runtime in 2010:
    title                                             runtime  budget      release_date
477                                 Ayirathil Oruvan  183      320000000.0 2010-01-14  
796                               Burnt by the Sun 2  181       45000000.0 2010-04-17  
610                                           Aurora  181              NaN 2010-05-14  
829           The Autobiography of Nicolae Ceausescu  180              NaN 2010-05-18  
66   Les Misérables in Concert: The 25th Anniversary  178             

In [33]:
# Define conversion rates for identified movies with high budgets
currency_adjustments = {
    'Enthiran': 0.013,  # INR to USD
    'Poetry': 0.00085,  # KRW to USD (approximate)
    'Goa': 0.013,       # INR to USD
    'Kites': 0.013      # INR to USD
}

# Iterate through the DataFrame to adjust budgets for known discrepancies
for index, row in imdb_2010_df.iterrows():
    title = row['title']
    if title in currency_adjustments:
        conversion_rate = currency_adjustments[title]
        imdb_2010_df.loc[index, 'budget'] = row['budget'] * conversion_rate

# Display the adjusted budgets for the specified movies
adjusted_titles = list(currency_adjustments.keys())
adjusted_budgets_df = imdb_2010_df[imdb_2010_df['title'].isin(adjusted_titles)][['title', 'budget', 'release_date']]
print("\nAdjusted Budgets for Identified Movies:")
print(adjusted_budgets_df)


Adjusted Budgets for Identified Movies:
    title      budget    release_date
284    Poetry     939.25 2010-05-13  
301  Enthiran  321100.00 2010-09-30  
335     Kites  101400.00 2010-05-21  
685       Goa  152100.00 2010-01-29  


In [34]:
%pip install pymongo
%pip install certifi


Note: you may need to restart the kernel to use updated packages.


*** Outliers Explanation
The analysis identified several notable outliers in terms of both budget and runtime for movies released in 2010.

*Budget Outliers
The highest budget outliers include Enthiran ($1.9 billion), Poetry ($1.3 billion), Goa ($900 million), and Kites ($600 million). These figures initially appeared inflated due to currency conversion issues. The dataset recorded the budgets in local currencies (e.g., INR for India, KRW for Korea), which resulted in inconsistencies. Howhwver in their own countries, Enthiran, Goa, and Kites showcase significant domestic investments in Indian cinema, while Poetry represents a substantial budget within the Korean market. At the same time, we can see low budjet Movie, such as Ceremony, casting Uma Turman and having low rating.

*Runtime Outliers
Among the runtime outliers, Ayirathil Oruvan (183 minutes), Burnt by the Sun 2 (181 minutes), and The Autobiography of Nicolae Ceausescu (180 minutes) are unusually long compared to the average film length in 2010. These longer runtimes are typically associated with epic narratives or documentaries, which often require more extended screen time to cover complex stories. The shortest runtimes are primarily for documentaries and short films, aligning with expectations.

Conclusion
The identified outliers reflect variability in global filmmaking practices, with longer runtimes aligning with specific genres or cultural storytelling traditions, while budget discrepancies were primarily due to initial mislabeling of currencies.

For improved accuracy, we need to check initial sourse of data and use better currency conversion methods and comprehensive metadata verification should be applied to prevent such discrepancies in future datasets. These measures could help distinguish genuine budget outliers from conversion errors, enhancing overall data integrity and analysis reliability.
